### ISSUE TITLE
Deterministically name wallet.dat file's

### ISSUE THREAD

**Issue by Fonta1n3** [cryptoadvance/specter-desktop#436]
Hey guys,

I am not familiar with your method of naming wallets so apologies if this issue is not relevant or feasible.

Since you are supporting the descriptor based wallet export/import aka Account Map I thought this proposal may also be something you are interested in.

In Fully Noded and Gordian Wallet I use the sha256 hash of the descriptor that the Account Map holds (the primary receive descriptor for the wallet) to deterministically name the wallet.dat file for the associated wallet on the app.

For example a Fully Noded wallet now looks like `FullyNoded-8268d751e68922956e6b3ee6ccbb469d18e8a158ee814aa47cb07d0259c1bbad.dat`

The main benefit occurs when a user: 

1. Switches devices
2. Connects new device to the same node 
3. Goes to recover their wallets on the new device 

At that point the app can use `bitcoin-cli listwalletdir` to check if the wallet already exists from `bitcoind` perspective by taking as input the Account Map's `descriptor`, getting the sha256 hash of it and cross checking all existing wallets to see if any match. 

If it does exist we can assume that it is safe to reuse that wallet as the chances of two hashes matching are nil. This results in much better UX as no additional wallet creation commands need to take place and more importantly no rescanning is required.
This is a hugely beneficial for pruned nodes as it will prevent users from having to "re-download the entire blockchain" every-time they want to recover a wallet on a different device.

This also fixes some edge cases that can arise if users use similar wallets across multiple apps (Specter, Fully Noded, Gordian Wallet) where one app may get neglected for a period of time which can result in the `pruneheight` being bypassed for an unloaded wallet and yet another reindex would be required which happened to me recently.

Just wanted to throw that out there to see if it could possibly fit in with how you are doing things.

Thanks for the great app!

**Comment by ben-kaufman** [comment]
Hi, thanks for sharing the idea.

So currently the way it works in Specter is that we have a separate folder called `specter_some_random_identifier` and under it, we save all the wallets folders with names based on the wallet name in Specter + a serial number if a wallet with that name already exists e.g `2of3_multisig`, `2of3_multisig_2`, `simple`, `simple_2` etc. (we convert the name to all lowercase and underscore instead of spaces).

The way FN does it sounds much better to me, as saving it by name is quite unnecessary since this isn't something users would normally care about, and since renaming wallet is possible and makes this way of naming somewhat confusing.
So I think the suggestion sounds good and very useful. I wonder though what would the best approach for interoperability be.

I guess what we can do is have Specter create wallets like FN as `Specter-descriptor_hash.dat`, then both FN and Specter (and others) on wallet creation would need to search for a wallet containing the same hash, regardless of the prefix, and use it if they found one. Does that sound right?

One comment/ question, the same descriptors can be written differently, like the order of xpubs in a `sortedmulti` descriptor, or derivation may be written slightly differently with the hardened symbol being possibly either `'`, `h`, or `H`. In order for this method to work, we need to make sure we use a single standardized format so we get matching hashes, is there something you already use for that?

I still want to make sure we can do this without making too much mess in our code or breaking backward compatibility (I'm pretty sure we can but just need to make sure).

**Comment by Fonta1n3** [comment]
> I guess what we can do is have Specter create wallets like FN as `Specter-descriptor_hash.dat`, then both FN and Specter (and others) on wallet creation would need to search for a wallet containing the same hash, regardless of the prefix, and use it if they found one. Does that sound right?

Yes that sounds correct, in FN I ignore the prefix altogether and just run a simple func which checks if the wallet name contains the hash or not:

```
if walletName.contains(theExpectedHash) {
    // reuse it
}
```
> 
> One comment/ question, the same descriptors can be written differently, like the order of xpubs in a `sortedmulti` descriptor, or derivation may be written slightly differently with the hardened symbol being possibly either `'`, `h`, or `H`. In order for this method to work, we need to make sure we use a single standardized format so we get matching hashes, is there something you already use for that?

Thats right, `sortedmulti` is the only chink here. In a practical use case if a user is running a pruned node and wants to avoid a potential 7 day download (raspi users), they would just have to compare their existing account map and make sure they add the xpubs in the same order. I think it is an acceptable tradeoff for now, unless there is a better idea.

As we discussed previously I am strictly relying on what we are putting in the Account Map, which is a descriptor using `h` instead of `'` and no checksum. Currently I just check for `'` and change it to `h`, I guess I should also check for `H` and change it to lowercase to avoid that.

> 
> I still want to make sure we can do this without making too much mess in our code or breaking backward compatibility (I'm pretty sure we can but just need to make sure).

Of course, take all the time you need! I have just used this to save myself having to reindex the entire blockchain and though it may be able to save others that headache too.

**Comment by stepansnigirev** [comment]
We should sort xpubs alphabetically then. The user doesn't have control over the order of cosigners in Specter, so we need to come up with the order ourselves.
I think the easiest is to sort alphabetically without derivation paths, so taking xpubs only.

**Comment by Fonta1n3** [comment]
> We should sort xpubs alphabetically then. The user doesn't have control over the order of cosigners in Specter, so we need to come up with the order ourselves.
> I think the easiest is to sort alphabetically without derivation paths, so taking xpubs only.

sounds good to me, will implement that on Fully Noded and Gordian too

**Comment by Fonta1n3** [comment]
Just an update on my end, i now always sort descriptor keys lexicographically so the wallet name is truly deterministic.

Ive got a user experiencing a weird bug I am trying to figure out... 

And it may have something to do with this issue.

By any chance has Specter adopted this hash technique for naming wallets?
